File: hw5/task1.py
from blackbox import BlackBox
import sys
from pyspark import SparkContext
import os
import time
import random
import binascii

def myhashs(s):
    p = int(1e5 + 3)
    user_int = int(binascii.hexlify(s.encode('utf8')), 16)
    result = []
    for a, b in hash_a_b:
        # Apply each hash function to the user ID
        hash_val = ((a * user_int + b) % p) % bit_len
        result.append(hash_val)
    return result

if __name__ == "__main__":

    sc = SparkContext(appName="Task1")

    #capture the parameter from linux command
    inputs = sys.argv[1]
    stream = int(sys.argv[2])
    asks = int(sys.argv[3])
    outputs = sys.argv[4]

    start_time = time.time()
    sc.setLogLevel("ERROR")
    #random.seed(553)
    bx = BlackBox()
    bit_len = 69997
    bit_ar = [0] * bit_len

    h_num = 30
    hash_a_b = [(random.randint(1, bit_len), random.randint(0, bit_len)) for i in range(h_num)]

    result_header = "Time,FPR\n"
    total = 0
    p_user = set()


    for i in range(asks):
        streamusers = bx.ask(inputs, stream)
        user_s = sc.parallelize(streamusers)
        user_h = user_s.map(lambda user: (user, myhashs(user)))
        results = user_h.collect()
        FP = 0

        for user, hashes in results:
            new = not all(bit_ar[h] for h in hashes)
            if new:
                #a new user, set the bits
                for h in hashes:
                    bit_ar[h] = 1
            else:
                if user not in p_user:
                # all bits are set, repeated situation
                    FP += 1
            p_user.add(user)
            total += 1

        fpr = FP / stream
        elem = [str(i), str(fpr)]
        result_header += ','.join(elem) + '\n'
    with open(outputs, 'w') as file:
        file.write(result_header)
    end_time = time.time()
    execution = end_time - start_time
    print(f"Duration: {execution}")
    sc.stop()

File: hw5/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import scala.util.Random
import scala.io.Source
import java.io._
import scala.collection.mutable.{Set, ArrayBuffer}
import java.math.BigInteger


object task1 {
  val bitLen = 69997
  val hNum = 30
  val random = new Random()
  val hashAB = Array.fill(hNum)((random.nextInt(bitLen) + 1, random.nextInt(bitLen + 1)))
  // Define the hash function
  def myHashes(s: String): Array[Int] = {
    val p: BigInt = BigInt(100003)// Close approximation to 1e5 + 3
    val userInt = BigInt(s.getBytes("UTF-8").map("%02x".format(_)).mkString, 16)
    hashAB.map { case (a, b) => ((BigInt(a) * userInt + b) % p % bitLen).toInt }
  }


  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setAppName("Task1")//.setMaster("local[*]")
    val sc = new SparkContext(conf)
    val inputs = args(0)
    val streamSize = args(1).toInt
    val numAsks = args(2).toInt
    val outputs = args(3)

    val bx = new Blackbox

    val start_time = System.currentTimeMillis()
    sc.setLogLevel("ERROR")

    val bitArray = Array.fill(bitLen)(false)
    var total = 0
    val pUser = Set.empty[String] // Previously seen users
    val resultHeader = new StringBuilder("Time,FPR\n")

    for (i <- 1 to numAsks) {
      var fp = 0

      val streamUsers = bx.ask(inputs, streamSize)
      val userS = sc.parallelize(streamUsers)
      val userH = userS.map(user => (user, myHashes(user))).collect()

      userH.foreach { case (user, hashes) =>
        val isNew = hashes.exists(hash => !bitArray(hash))
        if (isNew) hashes.foreach(hash => bitArray(hash) = true)
        else if (!pUser.contains(user)) fp += 1
        pUser.add(user)
        total += 1
      }

      val fpr = fp.toDouble / streamSize
      resultHeader.append(s"$i,$fpr\n")
    }

    val end_time = System.currentTimeMillis()
    val execution = (end_time - start_time) / 1000
    println(s"Duration: $execution")

    new PrintWriter(outputs) { write(resultHeader.toString); close() }
    sc.stop()
  }
}
